src/Shamir/Packed.hs

```diff
@@ -0,0 +1,306 @@
+{-|
+Module      : Shamir.Packed
+Description : Shamir packed secrets sharing scheme
+Copyright   : (c) 2019 Adjoint Inc.
+License     : MIT
+Maintainer  : "Adjoint Inc (info@adjoint.io)"
+-}
+
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE ParallelListComp #-}
+module Shamir.Packed
+  ( shareSecrets
+  , reconstructSecrets
+  , findNthRootOfUnity
+  , findNthPrimitiveRootOfUnity
+  , getRootOfUnity2
+  , getRootOfUnity3
+  , fft2
+  , fft3
+  , inverseDft2
+  , inverseDft3
+  , fftInterpolation2
+  , fftInterpolation3
+  , newtonInterpolation
+  , newtonEvaluate
+  , NewtonPolynomial(..)
+  , closestToPow2
+  , closestToPow3
+  ) where
+
+import Protolude hiding (quot)
+import Control.Error.Operator (assertM)
+import Control.Monad.Random (MonadRandom)
+import Data.Field.Galois (PrimeField(..), rnd, GaloisField, pow, char)
+import qualified Data.List as List
+import Data.Poly (VPoly, toPoly)
+import qualified Data.Vector as V
+
+-- | Polynomial represented as a coefficient vector, little-endian
+type CoeffVec f = [f]
+
+-- | Discrete Fourier transform. Can be interpreted as some polynomial
+-- evaluated at certain roots of unity. (In our case the length of
+-- these lists will be a power of two.)
+type DFT f = [f]
+
+--------------------
+-- Roots of unity --
+--------------------
+
+-- | Calculate roots of unity of powers of 2
+getRootOfUnity2 :: forall f. PrimeField f => Int -> f
+getRootOfUnity2 k
+  | 0 <= k     = 5^((char (witness :: f) - 1) `div` (2^k))
+  | otherwise  = panic "getRootOfUnity: No primitive root for given power of 2"
+
+-- | Calculate roots of unity of powers of 3
+getRootOfUnity3 :: forall f. PrimeField f => Int -> f
+getRootOfUnity3 k
+  | 0 <= k     = 5^((char (witness :: f) - 1) `div` (3^k))
+  | otherwise  = panic "getRootOfUnity: No primitive root for given power of 3"
+
+findNthRootOfUnity :: PrimeField f => (Int -> f) -> Int -> f
+findNthRootOfUnity getRootOfUnity n = go 2
+  where
+    go k = let root = getRootOfUnity k
+           in if root `pow` n == 1 then root else go (k + 1)
+
+-- | As prime fields are integral domains, every primitive nth root of unity
+-- is a principal root of unity. It is sufficient to verify that \alpha^{n} \neq 1 for 1 \leq k < n
+findNthPrimitiveRootOfUnity :: (PrimeField f, Enum f) => (Int -> f) -> Int -> f
+findNthPrimitiveRootOfUnity getRootOfUnity n = foldr f root [2..root]
+  where
+    root = findNthRootOfUnity getRootOfUnity n
+    f e lroot = if e `pow` n == 1 then e else lroot
+
+-------------------------------------
+-- Newton Polynomial Interpolation --
+-------------------------------------
+
+data NewtonPolynomial f = NewtonPolynomial
+  { npPoints :: [f]
+  , npCoeffs :: [f]
+  } deriving (Show, Eq)
+
+newtonInterpolation :: PrimeField f => [f] -> [f] -> NewtonPolynomial f
```

Comment:
Nice 🙂

Reply:
List lookups using ! are quite slow. This is the completely unoptimized version?

Reply:
Yes, we can do better. Let's use a map instead and add some benchmarks before merging